File: src/main/java/zxc/MrDrag0nXYT/nightJoiner/command/MainCommand.java
package zxc.MrDrag0nXYT.nightJoiner.command;

import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.configuration.file.YamlConfiguration;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import zxc.MrDrag0nXYT.nightJoiner.NightJoiner;
import zxc.MrDrag0nXYT.nightJoiner.util.Utilities;
import zxc.MrDrag0nXYT.nightJoiner.util.config.Config;
import zxc.MrDrag0nXYT.nightJoiner.util.config.Messages;
import zxc.MrDrag0nXYT.nightJoiner.util.database.DatabaseManager;
import zxc.MrDrag0nXYT.nightJoiner.util.database.DatabaseWorker;
import zxc.MrDrag0nXYT.nightJoiner.util.exception.UserRecordNotFound;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;

public class MainCommand implements CommandExecutor, TabCompleter {

    private final NightJoiner plugin;
    private Config config;
    private Messages messages;
    private DatabaseManager databaseManager;

    public MainCommand(NightJoiner plugin, Config config, Messages messages, DatabaseManager databaseManager) {
        this.plugin = plugin;
        this.config = config;
        this.messages = messages;
        this.databaseManager = databaseManager;
    }

    @Override
    public boolean onCommand(@NotNull CommandSender commandSender, @NotNull Command command, @NotNull String s, @NotNull String[] strings) {

        YamlConfiguration messagesConfig = this.messages.getConfig();

        if (strings.length == 0) {
            for (String string : messagesConfig.getStringList("nightjoiner.usage")) {
                commandSender.sendMessage(
                        Utilities.setColor(string)
                );
            }
            return true;
        }

        DatabaseWorker databaseWorker = databaseManager.getDatabaseWorker();

        switch (strings[0].toLowerCase()) {

            case "ban" -> {
                if (commandSender.hasPermission("nightjoiner.admin.ban")) {
                    if (strings.length == 2) {
                        try (Connection connection = databaseManager.getConnection()) {
                            databaseWorker.resetMessages(connection, Bukkit.getOfflinePlayer(strings[1]).getUniqueId(), strings[1]);
                            databaseWorker.setBlockStatus(connection, strings[1], 1);

                            for (String string : messagesConfig.getStringList("nightjoiner.banned")) {
                                commandSender.sendMessage(
                                        Utilities.setColor(
                                                string.replace("%player%", strings[1])
                                        )
                                );
                            }

                        } catch (SQLException e) {
                            e.printStackTrace();
                            for (String string : messagesConfig.getStringList("global.database-error")) {
                                commandSender.sendMessage(
                                        Utilities.setColor(string)
                                );
                            }

                        } catch (UserRecordNotFound e) {
                            for (String string : messagesConfig.getStringList("nightjoiner.player-not-found")) {
                                commandSender.sendMessage(
                                        Utilities.setColor(
                                                string.replace("%player%", strings[1])
                                        )
                                );
                            }
                        }
                    }

                } else {
                    for (String string : messagesConfig.getStringList("global.no-permission")) {
                        commandSender.sendMessage(
                                Utilities.setColor(string)
                        );
                    }
                }
            }

            case "unban" -> {
                if (commandSender.hasPermission("nightjoiner.admin.unban")) {
                    if (strings.length == 2) {
                        try (Connection connection = databaseManager.getConnection()) {
                            databaseWorker.setBlockStatus(connection, strings[1], 0);

                            for (String string : messagesConfig.getStringList("nightjoiner.unbanned")) {
                                commandSender.sendMessage(
                                        Utilities.setColor(
                                                string.replace("%player%", strings[1])
                                        )
                                );
                            }

                        } catch (SQLException e) {
                            e.printStackTrace();
                            for (String string : messagesConfig.getStringList("global.database-error")) {
                                commandSender.sendMessage(
                                        Utilities.setColor(string)
                                );
                            }

                        } catch (UserRecordNotFound e) {
                            for (String string : messagesConfig.getStringList("global.player-not-found")) {
                                commandSender.sendMessage(
                                        Utilities.setColor(
                                                string.replace("%player%", strings[1])
                                        )
                                );
                            }
                        }
                    }

                } else {
                    for (String string : messagesConfig.getStringList("global.no-permission")) {
                        commandSender.sendMessage(
                                Utilities.setColor(string)
                        );
                    }
                }
            }

            case "reload" -> {
                if (commandSender.hasPermission("nightjoiner.admin.reload")) {
                    plugin.reload();
                    for (String string : messagesConfig.getStringList("nightjoiner.reloaded")) {
                        commandSender.sendMessage(
                                Utilities.setColor(string)
                        );
                    }

                } else {
                    for (String string : messagesConfig.getStringList("global.no-permission")) {
                        commandSender.sendMessage(
                                Utilities.setColor(string)
                        );
                    }
                }
            }

            default -> {
                for (String string : messagesConfig.getStringList("nightjoiner.usage")) {
                    commandSender.sendMessage(
                            Utilities.setColor(string)
                    );
                }
            }
        }

        return true;
    }

    @Override
    public @Nullable List<String> onTabComplete(@NotNull CommandSender commandSender, @NotNull Command command, @NotNull String s, @NotNull String[] strings) {
        if (strings.length == 1) {
            return List.of("ban", "unban", "reload", "help");
        } else if (strings.length == 2) {
            return null;
        }

        return List.of();
    }
}

File: src/main/java/zxc/mrdrag0nxyt/nightJoiner/command/MainCommand.kt
package zxc.mrdrag0nxyt.nightJoiner.command

import net.kyori.adventure.text.TextReplacementConfig
import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import zxc.mrdrag0nxyt.nightJoiner.NightJoiner
import zxc.mrdrag0nxyt.nightJoiner.config.Messages
import zxc.mrdrag0nxyt.nightJoiner.database.DatabaseManager
import zxc.mrdrag0nxyt.nightJoiner.database.DatabaseWorker
import zxc.mrdrag0nxyt.nightJoiner.util.exception.UserRecordNotFound
import java.sql.SQLException
import java.util.*

class MainCommand(
    private val plugin: NightJoiner,
    private val messages: Messages,
    private val databaseManager: DatabaseManager
) :
    CommandExecutor, TabCompleter {
    override fun onCommand(commandSender: CommandSender, command: Command, s: String, strings: Array<String>): Boolean {
        if (strings.isEmpty()) return usage(commandSender)

        val databaseWorker = databaseManager.databaseWorker

        return when (strings[0].lowercase(Locale.getDefault())) {
            "reset" -> reset(commandSender, strings, databaseWorker)
            "ban" -> ban(commandSender, strings, databaseWorker)
            "unban" -> unban(commandSender, strings, databaseWorker)
            "reload" -> reload(commandSender)
            else -> usage(commandSender)
        }
    }

    override fun onTabComplete(
        commandSender: CommandSender,
        command: Command,
        s: String,
        strings: Array<String>
    ): List<String>? {
        if (strings.size == 1) {
            val complete = mutableListOf("help")

            if (commandSender.hasPermission("nightjoiner.admin.reload")) complete.add("reload")
            if (commandSender.hasPermission("nightjoiner.admin.reset")) complete.add("reset")
            if (commandSender.hasPermission("nightjoiner.admin.ban")) complete.add("ban")
            if (commandSender.hasPermission("nightjoiner.admin.unban")) complete.add("unban")

            return complete
        } else if (strings.size == 2) {
            return null
        }

        return listOf()
    }


    private fun usage(commandSender: CommandSender): Boolean {
        for (component in messages.mainUsage) {
            commandSender.sendMessage(component)
        }
        return false
    }

    private fun noPermission(commandSender: CommandSender): Boolean {
        for (component in messages.globalNoPermission) {
            commandSender.sendMessage(component)
        }
        return false
    }

    private fun reset(
        commandSender: CommandSender,
        strings: Array<String>,
        databaseWorker: DatabaseWorker?
    ): Boolean {
        if (!commandSender.hasPermission("nightjoiner.admin.reset")) return noPermission(commandSender)

        if (strings.size >= 2) {
            plugin.server.scheduler.runTaskAsynchronously(plugin, Runnable {
                try {
                    databaseManager.getConnection().use { connection ->
                        databaseWorker?.resetMessages(
                            connection!!, Bukkit.getOfflinePlayer(strings[1]).uniqueId,
                            strings[1]
                        )
                        val textReplacementConfig = TextReplacementConfig.builder()
                            .match("%player%")
                            .replacement(strings[1])
                            .build()

                        for (component in messages.mainTargetReset) {
                            commandSender.sendMessage(
                                component.replaceText(textReplacementConfig)
                            )
                        }
                    }
                } catch (e: SQLException) {
                    e.printStackTrace()
                    for (component in messages.globalDatabaseError) {
                        commandSender.sendMessage(component)
                    }
                } catch (e: UserRecordNotFound) {
                    val textReplacementConfig = TextReplacementConfig.builder()
                        .match("%player%")
                        .replacement(strings[1])
                        .build()

                    for (component in messages.mainTargetNotFound) {
                        commandSender.sendMessage(
                            component.replaceText(textReplacementConfig)
                        )
                    }
                }
            })
            return true

        } else return usage(commandSender)
    }

    private fun ban(
        commandSender: CommandSender,
        strings: Array<String>,
        databaseWorker: DatabaseWorker?
    ): Boolean {
        if (!commandSender.hasPermission("nightjoiner.admin.ban")) return noPermission(commandSender)

        if (strings.size >= 2) {
            plugin.server.scheduler.runTaskAsynchronously(plugin, Runnable {
                try {
                    databaseManager.getConnection().use { connection ->
                        databaseWorker?.resetMessages(
                            connection!!, Bukkit.getOfflinePlayer(strings[1]).uniqueId,
                            strings[1]
                        )
                        databaseWorker?.setBlockStatus(connection!!, strings[1], 1)
                        val textReplacementConfig = TextReplacementConfig.builder()
                            .match("%player%")
                            .replacement(strings[1])
                            .build()

                        for (component in messages.mainTargetBanned) {
                            commandSender.sendMessage(
                                component.replaceText(textReplacementConfig)
                            )
                        }
                    }
                } catch (e: SQLException) {
                    e.printStackTrace()
                    for (string in messages.globalDatabaseError) {
                        commandSender.sendMessage(string)
                    }
                } catch (e: UserRecordNotFound) {
                    val textReplacementConfig = TextReplacementConfig.builder()
                        .match("%player%")
                        .replacement(strings[1])
                        .build()

                    for (component in messages.mainTargetNotFound) {
                        commandSender.sendMessage(
                            component.replaceText(textReplacementConfig)
                        )
                    }
                }
            })
            return true

        } else return usage(commandSender)
    }

    private fun unban(
        commandSender: CommandSender,
        strings: Array<String>,
        databaseWorker: DatabaseWorker?
    ): Boolean {
        if (!commandSender.hasPermission("nightjoiner.admin.unban")) return noPermission(commandSender)

        if (strings.size >= 2) {
            plugin.server.scheduler.runTaskAsynchronously(plugin, Runnable {
                try {
                    databaseManager.getConnection().use { connection ->
                        databaseWorker!!.setBlockStatus(connection!!, strings[1], 0)
                        val textReplacementConfig = TextReplacementConfig.builder()
                            .match("%player%")
                            .replacement(strings[1])
                            .build()

                        for (component in messages.mainTargetUnbanned) {
                            commandSender.sendMessage(
                                component.replaceText(textReplacementConfig)
                            )
                        }
                    }
                } catch (e: SQLException) {
                    e.printStackTrace()
                    for (string in messages.globalDatabaseError) {
                        commandSender.sendMessage(string)
                    }
                } catch (e: UserRecordNotFound) {
                    val textReplacementConfig = TextReplacementConfig.builder()
                        .match("%player%")
                        .replacement(strings[1])
                        .build()

                    for (component in messages.mainTargetNotFound) {
                        commandSender.sendMessage(
                            component.replaceText(textReplacementConfig)
                        )
                    }
                }
            })
            return true

        } else return usage(commandSender)
    }

    private fun reload(commandSender: CommandSender): Boolean {
        if (!commandSender.hasPermission("nightjoiner.admin.reload")) return noPermission(commandSender)

        plugin.reload()
        for (string in messages.mainReloaded) {
            commandSender.sendMessage(string)
        }
        return true
    }
}
